Added support to Parse ASYNC function

mvvsmk opened the pull request:
Fixes issue #2650 
The parser now parses ASYNC functions.
gcc/rust/ChangeLog:

	* parse/rust-parse-impl.h (Parser::parse_item):Added ASYNC case.
	(Parser::parse_vis_item):Added a switch case to handle ASYNC.

gcc/testsuite/ChangeLog:

	* rust/compile/issue-2650.rs: New test.

Signed-off-by : M V V S Manoj Kumar <mvvsmanojkumar@gmail.com>

dkm commented:
Typo in commit log (handle).
Also, you probably want to refer to #2650 instead of #2560

mvvsmk commented:
Thanks for pointing out the typo and the link error. I have reworded my commit to reflect it.

CohenArthur commented on gcc/rust/parse/rust-parse-impl.h:
```diff
@@ -1385,6 +1387,25 @@ Parser<ManagedTokenSource>::parse_vis_item (AST::AttrVec outer_attrs)
 	  lexer.skip_token (1); // TODO: is this right thing to do?
 	  return nullptr;
 	}
+    case ASYNC:
+      t = lexer.peek_token (1);
+
+      switch (t->get_id ())
+	{
+	case UNSAFE:
+	case FN_TOK:
+	  return parse_function (std::move (vis), std::move (outer_attrs));
+
+	default:
+	  add_error (
+	    Error (t->get_locus (),
+		   "unexpected token %qs in some sort of async production",
+		   t->get_token_description ()));
+
+	  lexer.skip_token (1);
+	  return nullptr;
+	}
+
```
async is not allowed if the Rust edition is set to 2015, which is the default. I think keeping the parser like that is fine, but we should add that to the checks in the AST validation pass

P-E-P replied:
Rustc error out on `async` during parsing. We probably want to do the same to ensure macro compatibility.

mvvsmk replied:
Could you tell me where or how I can find which version of Rust is set ? I know the magic happens in the lang.opt file, but how do we know what parameters are passed?

P-E-P replied:
AFAIK there is a singleton in the rust session manager.

CohenArthur commented on gcc/testsuite/rust/compile/issue-2650.rs:
```diff
@@ -0,0 +1,3 @@
+pub async fn a() -> u32 {
+    1
+}
```
so this test should eventually fail, unless we are in an edition >= 2018. so please do the following:
```suggestion
// { dg-additional-options "-frust-edition=2018" }

pub async fn a() -> u32 {
    1
}
```

mvvsmk replied:
I see, I'll do that. I am not familiar with dejagnu , but wouldn't it be "-frust-edition>=2018" or does it handle it on its own?

P-E-P replied:
This dejagnu directive pass the argument to the compiler, this means the test code will be compiled using gccrs with the flag `-frust-edition=2018` hence why we're not setting `>=`, because we're setting the exact rust edition, we're not checking against all editions greater than 2018.

mvvsmk replied:
Oh , I got it now, thanks for the help :D

CohenArthur commented:
example of the error if the Rust edition is set to 2015: https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=9696f6f21bd504217a3b9dab4947f507

P-E-P commented on gcc/rust/parse/rust-parse-impl.h:
```diff
@@ -1385,6 +1387,25 @@ Parser<ManagedTokenSource>::parse_vis_item (AST::AttrVec outer_attrs)
 	  lexer.skip_token (1); // TODO: is this right thing to do?
 	  return nullptr;
 	}
+    case ASYNC:
```
I feel like we could have a new function here, something along `parse_async_item`, we already have `parse_const_item`.

mvvsmk replied:
Understood, I'll make this change.

mvvsmk replied:
Hey @P-E-P , in our parser we have a function called `parse_async_block`, which has just has a defination and no caller in the entire codebase, would you like me to make a new `parse_async_item` function or just ammend my code to `parse_async_block`

P-E-P replied:
I think `parse_async_block` was made exclusively for async block (`async {}`) you should probably create a new function instead of using this one.

mvvsmk replied:
Oh ok, got it :)

P-E-P reviewed (approved):
I'll be happy to merge this once the edition thing will be sorted out.

mvvsmk commented:
I have made the requested changes, if I missed anything do let me know :). 

I also added another test that fails, so we have both cases covered. I also noticed that our `parse_async_block` has no such check, would you like me to add it in this PR itself or make a separate one?

also a minor change but in rustc error says :
```
`async fn` is not permitted in Rust 2015
```
but the cpp11 compiler just kept giving me warnings about the " ` " symbol don't know what that was about, spent half the day figuring it out but it would still give me a warning no matter what I did so, changed it to :
```
'async fn' is not permitted in Rust 2015
```
if you guys got any hints about it do let me know :)

P-E-P commented:
> I have made the requested changes, if I missed anything do let me know :).
> 
> I also added another test that fails, so we have both cases covered. I also noticed that our `parse_async_block` has no such check, would you like me to add it in this PR itself or make a separate one?
> 
> also a minor change but in rustc error says :
> 
> ```
> `async fn` is not permitted in Rust 2015
> ```
> 
> but the cpp11 compiler just kept giving me warnings about the " ` " symbol don't know what that was about, spent half the day figuring it out but it would still give me a warning no matter what I did so, changed it to :
> 
> ```
> 'async fn' is not permitted in Rust 2015
> ```
> 
> if you guys got any hints about it do let me know :)

That's because of quoted string. If the quoted thing lives in a variable you might use `%qX` with X being the correct format argument. For `*char` that would be `%qs`. Otherwise you can use `%<` and `%>`.

P-E-P commented on gcc/rust/parse/rust-parse-impl.h:
```diff
@@ -1427,6 +1434,40 @@ Parser<ManagedTokenSource>::parse_vis_item (AST::AttrVec outer_attrs)
+      == CompileOptions::Edition::E2015)
+    {
+      add_error (Error (Error::Kind::Err, t->get_locus (), ErrorCode::E0670,
+			" %<`async fn`%> is not permitted in Rust 2015"));
```
`%<` already puts a ` in your output text, you do not need to keep those `

mvvsmk replied:
Done :D

P-E-P commented on gcc/testsuite/rust/compile/issue-2650-2.rs:
```diff
@@ -0,0 +1,6 @@
+// { dg-additional-options "-frust-edition=2015" }
+
+// { dg-excess-errors "'async fn' is not permitted in Rust 2015" }
```
You won't be able to match the backtick token using dejaGNU because AFAIK it is based on the locale. User a dot `.` to match any character instead.

mvvsmk commented on gcc/testsuite/rust/compile/issue-2650-2.rs:
```diff
@@ -0,0 +1,6 @@
+// { dg-additional-options "-frust-edition=2015" }
+
+// { dg-excess-errors "'async fn' is not permitted in Rust 2015" }
```
Sure, I'll get it done.

CohenArthur reviewed (approved):
LGTM :) nice work!

CohenArthur commented on gcc/rust/parse/rust-parse-impl.h:
```diff
@@ -1427,6 +1434,40 @@ Parser<ManagedTokenSource>::parse_vis_item (AST::AttrVec outer_attrs)
+
+    default:
+      add_error (Error (t->get_locus (),
+			"unexpected token %qs in some sort of async production",
```
once this is addressed we should merge the PR

mvvsmk replied:
I have made this change but I was experimeting a bit with rustc and found this error code 
```
error: expected one of `extern`, `fn`, or `unsafe`, found keyword `const`
 --> src/lib.rs:1:7
  |
1 | async const fn foo() {
  | ------^^^^^
  | |     |
  | |     expected one of `extern`, `fn`, or `unsafe`
  | help: `const` must come before `async`: `const async`
  ```
  would you like me to go with that and make the relavent changes by adding extern to the next token list and change the error message to the one emmited by rustc?
  
  this is the code I was testing : https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=38fc87fda8f9debc717354d3af2ef863

P-E-P replied:
@mvvsmk I think it would fit better in another PR

mvvsmk commented:
@P-E-P Could you add this back to the merge queue when you get time, there was a merge conflict so it was removed from the queue.
